aten/src/ATen/native/cpu/HistogramKernel.cpp

```diff
@@ -81,138 +91,170 @@ void histogram_cpu_contiguous(Tensor& hist, const Tensor& bin_edges,
             }
 
-            // In the unweighted case, the default weight is 1
-            input_t wt = accessor_wt.has_value() ? accessor_wt.value()[i] : static_cast<input_t>(1);
-            data_out_local[pos] += wt;
+            if (!skip_elt) {
+                // In the unweighted case, the default weight is 1
+                input_t wt = accessor_wt.has_value() ? accessor_wt.value()[i] : static_cast<input_t>(1);
+
+                input_t cur = hist_local.index(indices).item<input_t>();
+                hist_local.index_put_(indices, cur + wt);
+            }
         }
 
+
         // Locks and updates the common output
-        const std::lock_guard<std::mutex> lock(data_out_mutex);
-        for (int64_t i = 0; i < numel_be - 1; i++) {
-            data_out[i] += data_out_local[i];
-        }
+        const std::lock_guard<std::mutex> lock(hist_mutex);
+        hist.add_(hist_local);
     });
 }
 
 /* Some pre- and post- processing steps for the main algorithm.
  * Initializes hist to 0, calls into the main algorithm, and normalizes output if necessary.
  */
 template<BIN_SELECTION_ALGORITHM bin_algorithm>
-void histogram_out_cpu_template(const Tensor& self, const c10::optional<Tensor>& weight, bool density,
-        Tensor& hist, const Tensor& bin_edges) {
+void histogramdd_out_cpu_template(const Tensor& self, const c10::optional<Tensor>& weight, bool density,
+        Tensor& hist, const TensorList& bin_edges) {
     hist.fill_(0);
 
-    const int64_t numel_in = self.numel();
-    const Tensor reshaped_input = self.reshape({numel_in});
+    const int64_t D = self.size(-1);
+    const int64_t N = self.numel() / D;
+
+    const Tensor reshaped_input = self.reshape({N, D});
 
     const auto reshaped_weight = weight.has_value()
-            ? c10::optional<Tensor>(weight.value().reshape({numel_in}))
+            ? c10::optional<Tensor>(weight.value().reshape({N}))
             : c10::optional<Tensor>();
 
+    std::vector<Tensor> bin_edges_contig(bin_edges.size());
+    for (size_t dim = 0; dim < bin_edges_contig.size(); dim++) {
+        bin_edges_contig[dim] = bin_edges[dim].contiguous();
+    }
+
     AT_DISPATCH_FLOATING_TYPES(self.scalar_type(), "histogram_cpu", [&]() {
-        histogram_cpu_contiguous<scalar_t, bin_algorithm>(
-                hist, bin_edges.contiguous(), reshaped_input, reshaped_weight);
+        histogramdd_cpu_contiguous<scalar_t, bin_algorithm>(
+                hist, bin_edges_contig, reshaped_input, reshaped_weight);
     });
 
     // Converts the bin totals to a probability density function
     if (density) {
-        auto bin_widths = bin_edges.diff();
         auto hist_sum = hist.sum().item();
-        hist.div_(bin_widths);
+
+        std::vector<int64_t> shape(D, 1), reps(hist.sizes().vec());
+
+        for (int64_t dim = 0; dim < D; dim++) {
+            std::swap(shape[dim], reps[dim]);
+            hist.div_(bin_edges[dim].diff().reshape(shape).repeat(reps));
```

Comment:
Explore computing product of the bin_edge diffs instead, add comments either way.